### PR TITLE
[kcalcore] Calculate LocalTime from UTC DTSTART when parsing phases. Contributes to MER#1011

### PR DESCRIPTION
**Pull request by chriadam**
This commit ensures that the interval within which transitions are
determined (from the RRULE) is fully-specified, otherwise the interval
is considered null and thus no transitions are deemed to occur.

Previously, the start of the interval is set to the DTSTART value,
but if that DTSTART value is specified in UTC format, this fails due
to a bad conversion to local time.

This commit ensures that the start of the interval is calculated
(in local time) from the (correctly parsed) UTC DTSTART value.

Contributes to MER#1011

### CODE REVIEW

**Comment by chriadam** [comment]
@pvuorela @pgerdt @blammit

**Comment by pvuorela** [comment]
Trying this with test I created earlier for the Berlin case, I seem to be getting timezone utc+1 for event in april even though it should be +2. Also test app from lenggi still displays different times.

Not sure I got the libical problem you spotted yet, but in case there's a bug, maybe we should upgrade to newer version? Current 0.48 is from 2011.

**Comment by chriadam** [comment]
Er, you're right, I was using a different TIMEZONE definition in my test for some reason... I'll investigate further, sorry for the noise.

**Comment by chriadam** [comment]
Ok, there were two problems.  I've added a new commit which resolves the second (original) problem to this PR.

**Comment by chriadam** [comment]
Here is my complete test application for the identified and fixed issues:

    #include <QCoreApplication>
    #include <QTimer>
    #include <QtDebug>

    // kcalcore
    #include <extendedcalendar.h>
    #include <extendedstorage.h>
    #include <notebook.h>
    #include <incidence.h>
    #include <icalformat.h>
    #include <memorycalendar.h>

    // Here we have an event which we import.
    // The vcalendar also specifies the timezone data for the event.
    // Assumes the current system timezone is set to Europe/Berlin.
    int testOne()
    {
        QByteArray icalData =
            "BEGIN:VCALENDAR\r\n"
            "VERSION:2.0\r\n"
            "PRODID:-//EGroupware//NONSGML EGroupware Calendar 1.9.004//DE\r\n"
            "BEGIN:VTIMEZONE\r\n"
            "TZID:Europe/Berlin\r\n"
            "X-LIC-LOCATION:Europe/Berlin\r\n"
            "BEGIN:DAYLIGHT\r\n"
            "TZOFFSETFROM:+0100\r\n"
            "TZOFFSETTO:+0200\r\n"
            "TZNAME:CEST\r\n"
            "DTSTART:19700329T020000\r\n"
            "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3\r\n"
            "END:DAYLIGHT\r\n"
            "BEGIN:STANDARD\r\n"
            "TZOFFSETFROM:+0200\r\n"
            "TZOFFSETTO:+0100\r\n"
            "TZNAME:CET\r\n"
            "DTSTART:19701025T030000\r\n"
            "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10\r\n"
            "END:STANDARD\r\n"
            "END:VTIMEZONE\r\n"
            "BEGIN:VEVENT\r\n"
            "CLASS:PUBLIC\r\n"
            "SUMMARY:REDACTED\r\n"
            "DTSTART;TZID=Europe/Berlin:20150402T130000\r\n"
            "DTEND;TZID=Europe/Berlin:20150402T140000\r\n"
            "PRIORITY:5\r\n"
            "TRANSP:OPAQUE\r\n"
            "CATEGORIES:REDACTED\r\n"
            "UID:calendar-42066-f2bd5fa3746e5d6bdbfac09ac22cc8c2\r\n"
            "STATUS:CONFIRMED\r\n"
            "CREATED:20150330T152325Z\r\n"
            "LAST-MODIFIED:20150330T152325Z\r\n"
            "DTSTAMP:20150330T154508Z\r\n"
            "END:VEVENT\r\n"
            "END:VCALENDAR\r\n\r\n";

        KCalCore::ICalFormat iCalFormat;
        KCalCore::MemoryCalendar::Ptr calendar(new KCalCore::MemoryCalendar(KDateTime::UTC));
        if (!iCalFormat.fromString(calendar, icalData)) {
            qWarning() << "ical import failed!";
            return 1;
        }
        KCalCore::Event::List events = calendar->events();
        KCalCore::Event::Ptr evt;
        if (events.size()) {
            evt = events.first();
        }

        QString dtStartStringUtc = evt->dtStart().toString();
        QString dtEndStringUtc = evt->dtEnd().toString();
        qWarning() << "event:" << dtStartStringUtc << "->" << dtEndStringUtc;
        // expect: event: "2015-04-02T13:00:00+02:00" -> "2015-04-02T14:00:00+02:00"
        calendar->close();

        if (dtStartStringUtc == QStringLiteral("2015-04-02T13:00:00+02:00")
                && dtEndStringUtc == QStringLiteral("2015-04-02T14:00:00+02:00")) {
            return 0; // success
        } else {
            return 1; // failure.
        }
    }

    // Here we just have a timezone specification in the vcalendar.
    // We then use this timezone specification to adjust a particular
    // datetime.  The DTSTART in the timezone specification previously
    // caused problems when defining the transitions determination interval.
    // Assumes the current system timezone is set to Europe/Berlin.
    int testTwo()
    {
        QByteArray icalData =
            "BEGIN:VCALENDAR\r\n"
            "VERSION:2.0\r\n"
            "PRODID:-//EGroupware//NONSGML EGroupware Calendar 1.9.004//DE\r\n"
            "BEGIN:VTIMEZONE\r\n"
            "TZID:Europe/Berlin\r\n"
            "BEGIN:DAYLIGHT\r\n"
            "TZNAME:CEST\r\n"
            "TZOFFSETFROM:+0100\r\n"
            "TZOFFSETTO:+0200\r\n"
            "DTSTART:19810329T020000\r\n"
            "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3\r\n"
            "END:DAYLIGHT\r\n"
            "BEGIN:STANDARD\r\n"
            "TZNAME:CET\r\n"
            "TZOFFSETFROM:+0200\r\n"
            "TZOFFSETTO:+0100\r\n"
            "DTSTART:19800928T030000\r\n"
            "RRULE:FREQ=YEARLY;COUNT=16;BYDAY=-1SU;BYMONTH=9\r\n"
            "END:STANDARD\r\n"
            "BEGIN:STANDARD\r\n"
            "TZNAME:CET\r\n"
            "TZOFFSETFROM:+0200\r\n"
            "TZOFFSETTO:+0100\r\n"
            "DTSTART:19961027T030000\r\n"
            "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10\r\n"
            "END:STANDARD\r\n"
            "END:VTIMEZONE\r\n"
            "END:VCALENDAR\r\n\r\n";

        KCalCore::ICalFormat iCalFormat;
        KCalCore::MemoryCalendar::Ptr calendar = KCalCore::MemoryCalendar::Ptr(new KCalCore::MemoryCalendar(KDateTime::Spec::UTC()));
        if (!iCalFormat.fromString(calendar, icalData)) {
            qWarning() << "ical import failed!";
            return 1;
        }

        KCalCore::ICalTimeZoneSource tzsource;
        KCalCore::ICalTimeZone berlinSystem = tzsource.standardZone( "Europe/Berlin" );
        KCalCore::ICalTimeZones *tzlist = calendar->timeZones();
        KCalCore::ICalTimeZone berlinParsed = tzlist->zone("Europe/Berlin");
        KDateTime::Spec timeSpecSystem = KDateTime::Spec( berlinSystem );
        KDateTime::Spec timeSpecParsed = KDateTime::Spec( berlinParsed );

        QDateTime myTime = QDateTime( QDate( 2015, 05, 01 ), QTime( 14, 00, 00 ), Qt::UTC);

        qWarning() << "system timezone utc offset:" << timeSpecSystem.utcOffset();
        qWarning() << "parsed timezone utc offset:" << timeSpecParsed.utcOffset();
        qWarning() << "equivalent? " << timeSpecSystem.equivalentTo(timeSpecParsed);

        QDateTime systemTime = timeSpecSystem.timeZone().toZoneTime(myTime);
        QDateTime parsedTime = timeSpecParsed.timeZone().toZoneTime(myTime);

        qWarning() << "system timezone for" << myTime.toString() << "gets you" << systemTime.toString(Qt::ISODate);
        qWarning() << "parsed timezone for" << myTime.toString() << "gets you" << parsedTime.toString(Qt::ISODate);
        calendar->close();

        // should be equivalent in UTC.
        if (systemTime.toString() == parsedTime.toString()) {
            return 0; // success
        } else {
            return 1;
        }
    }

    // Here we just have a timezone specification in the vcalendar.
    // We then use this timezone specification to adjust a particular
    // datetime.  The timezone specification previously caused individual
    // transition dates to be deemed invalid by the system, as their
    // timespec had conflicted with the system timespec.
    // Assumes the current system timezone is set to Europe/Berlin.
    int testThree()
    {
        QByteArray icalData =
            "BEGIN:VCALENDAR\r\n"
            "VERSION:2.0\r\n"
            "PRODID:-//EGroupware//NONSGML EGroupware Calendar 1.9.004//DE\r\n"
            "BEGIN:VTIMEZONE\r\n"
            "TZID:Europe/Berlin\r\n"
            "X-LIC-LOCATION:Europe/Berlin\r\n"
            "BEGIN:DAYLIGHT\r\n"
            "TZOFFSETFROM:+0100\r\n"
            "TZOFFSETTO:+0200\r\n"
            "TZNAME:CEST\r\n"
            "DTSTART:19700329T020000\r\n"
            "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3\r\n"
            "END:DAYLIGHT\r\n"
            "BEGIN:STANDARD\r\n"
            "TZOFFSETFROM:+0200\r\n"
            "TZOFFSETTO:+0100\r\n"
            "TZNAME:CET\r\n"
            "DTSTART:19701025T030000\r\n"
            "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=10\r\n"
            "END:STANDARD\r\n"
            "END:VTIMEZONE\r\n"
            "END:VCALENDAR\r\n\r\n";

        KCalCore::ICalFormat iCalFormat;
        KCalCore::MemoryCalendar::Ptr calendar = KCalCore::MemoryCalendar::Ptr(new KCalCore::MemoryCalendar(KDateTime::Spec::UTC()));
        if (!iCalFormat.fromString(calendar, icalData)) {
            qWarning() << "ical import failed!";
            return 1;
        }

        KCalCore::ICalTimeZoneSource tzsource;
        KCalCore::ICalTimeZone berlinSystem = tzsource.standardZone( "Europe/Berlin" );
        KCalCore::ICalTimeZones *tzlist = calendar->timeZones();
        KCalCore::ICalTimeZone berlinParsed = tzlist->zone("Europe/Berlin");
        KDateTime::Spec timeSpecSystem = KDateTime::Spec( berlinSystem );
        KDateTime::Spec timeSpecParsed = KDateTime::Spec( berlinParsed );

        QDateTime myTime = QDateTime( QDate( 2015, 05, 01 ), QTime( 14, 00, 00 ), Qt::UTC);

        qWarning() << "system timezone utc offset:" << timeSpecSystem.utcOffset();
        qWarning() << "parsed timezone utc offset:" << timeSpecParsed.utcOffset();
        qWarning() << "equivalent? " << timeSpecSystem.equivalentTo(timeSpecParsed);

        QDateTime systemTime = timeSpecSystem.timeZone().toZoneTime(myTime);
        QDateTime parsedTime = timeSpecParsed.timeZone().toZoneTime(myTime);

        qWarning() << "system timezone for" << myTime.toString() << "gets you" << systemTime.toString(Qt::ISODate);
        qWarning() << "parsed timezone for" << myTime.toString() << "gets you" << parsedTime.toString(Qt::ISODate);
        calendar->close();

        // should be equivalent in UTC.
        if (systemTime.toString() == parsedTime.toString()) {
            return 0; // success
        } else {
            return 1;
        }
    }

    int main(int argc, char *argv[])
    {
        QCoreApplication app(argc, argv);
        QTimer timer;
        timer.setSingleShot(true);
        timer.setInterval(2000);
        QObject::connect(&timer, &QTimer::timeout, &app, &QCoreApplication::quit);
        timer.start();

        if (testOne() != 0) {
            qWarning() << "test one failed!";
        } else {
            qWarning() << "test one succeeded!";
        }

        if (testTwo() != 0) {
            qWarning() << "test two failed!";
        } else {
            qWarning() << "test two succeeded!";
        }

        if (testThree() != 0) {
            qWarning() << "test three failed!";
        } else {
            qWarning() << "test three succeeded!";
        }

        return app.exec();
    }

All three tests fail with current kcalcore from OBS, and succeed with this PR applied, for me, with device set to Europe/Berlin timezone.

File: kcalcore/icaltimezones.cpp [hunk 1]
```diff
@@ -1313,10 +1313,10 @@ QList<QDateTime> ICalTimeZoneSourcePrivate::parsePhase( icalcomponent *c,
   }
 
   // Convert DTSTART to QDateTime, and from local time to UTC
-  const QDateTime localStart = toQDateTime( dtstart );   // local time
   dtstart.second -= prevOffset;
   dtstart.is_utc = 1;
   const QDateTime utcStart = toQDateTime( icaltime_normalize( dtstart ) );   // UTC
+  const QDateTime localStart = utcStart.toLocalTime();  // local time
```
Comment (chriadam) [inline kcalcore/icaltimezones.cpp]: Hrm, this may convert the utcStart into localStart where localStart is in the current SYSTEM local time; which may not be what is required here... the original code may have been intended to find the localStart in the timezone specified by the VCALENDAR entry VTIMEZONE section, perhaps?

Comment (pvuorela) [inline kcalcore/icaltimezones.cpp]: Yeah, was thinking the same. What's the test case in your example this fixes? And how about that libical update?

File: kcalcore/recurrencerule.cpp [hunk 2]
```diff
@@ -623,6 +623,16 @@ void Constraint::appendDateTime( const QDate &date, const QTime &time,
...
 {
   KDateTime dt( date, time, timespec );
+  KDateTime nonTimespecDt( date, time, KDateTime::Spec::UTC() );
+  if ( !dt.isValid() && nonTimespecDt.isValid() && timespec.utcOffset() == 0 ) {
+    // This is to fix an issue where CalDAV servers can provide a VTIMEZONE
+    // definition for CEST (eg "Europe/Berlin") with a DAYLIGHT component like:
+    // "DTSTART:19700329T020000\r\n"
+    // "RRULE:FREQ=YEARLY;BYDAY=-1SU;BYMONTH=3\r\n"
+    // This will fail after 1981 since the system definition differs, resulting
+    // in an invalid KDateTime due to the "invalid" timespec.
```
Comment (pvuorela) [inline kcalcore/recurrencerule.cpp]: So, the member variable timespec will have SpecType of Invalid? Doesn't sound like something to be trusted for utfOffset.

File: kcalcore/icaltimezones.cpp (hunk 1 above)
Comment (chriadam) [inline kcalcore/icaltimezones.cpp]: This fixes testTwo() from the test app, for me.  Well, it fixed testTwo() for me without the second change applied already... I wonder if this change is unnecessary if we just apply the second change (ie, maybe the second change fixes all issues, even without this change?)  I'll test this today.

Regarding libical upgrade - this is a nice idea, as I believe that the current version of libical we use has a memory leak in it, too, related to timezones IIRC.  But I haven't tested.  I don't have write access to mer-core/libical.git though, but I will try updating to latest upstream in a local branch and see if updating helps, also.

File: kcalcore/recurrencerule.cpp (hunk 2 above)
Comment (chriadam) [inline kcalcore/recurrencerule.cpp]: I didn't print out the spectype tbh, but I suspect that it was Invalid, yes.  All I know is that the KDateTime becomes invalid if you use that member variable timespec, for dates AFTER 1981, while using the VTIMEZONE definition provided in testThree() (and also testOne() I assume).

But interestingly, in my tests, the utcOffset of the member variable timespec is ALWAYS zero!  The "time" variable passed to the function is already adjusted for the UTC offset defined by the VTIMEZONE definition, by that point.

In short, I suspect (but am not 100% certain) that we could _always_ use the system UTC timespec when creating the dt here.  But, I thought it safer to only "fall back" to that behaviour, obviously.

File: kcalcore/icaltimezones.cpp (hunk 1 above)
Comment (chriadam) [inline kcalcore/icaltimezones.cpp]: Interesting... testTwo() succeeds (along with testOne() and testThree()) if we have the second patch but not the first patch included.  So it looks like the first patch is not necessary, now.

However, if we put a line like:
    qWarning() << "calculating transitions within interval:" << klocalStart.toString() << "->" << maxTime.toString();
just above the line:
    const DateTimeList dts = r.timesInInterval( klocalStart, maxTime );
then you can see without the first patch, it prints:
    calculating transitions within interval: "" -> "2035-05-21T00:00:00"
whereas with the first patch included, it prints:
    calculating transitions within interval: "1981-03-29T03:00:00" -> "2035-05-21T00:00:00"

Nonetheless, given the potential for that first patch to cause other issues, perhaps we should not do that one, and just apply the second patch in this PR?

Comment (chriadam) [inline kcalcore/icaltimezones.cpp]: I built an upgraded libical, but I can't install it without updating a bunch of other packages which require it:

	libical.so.0 is needed by (installed) nemo-qml-plugin-calendar-qt5-lightweight-0.1.4-1.armv7hl
	libical.so.0 is needed by (installed) nemo-qml-plugin-calendar-qt5-0.1.4-1.armv7hl
	libical.so.0 is needed by (installed) mkcal-qt5-0.3.27-1.16.1.armv7hl
	libical.so.0 is needed by (installed) kcalcore-qt5-4.10.2-1.armv7hl
	libicalss.so.0 is needed by (installed) nemo-qml-plugin-calendar-qt5-lightweight-0.1.4-1.armv7hl
	libicalss.so.0 is needed by (installed) nemo-qml-plugin-calendar-qt5-0.1.4-1.armv7hl
	libicalss.so.0 is needed by (installed) mkcal-qt5-0.3.27-1.16.1.armv7hl
	libicalss.so.0 is needed by (installed) kcalcore-qt5-4.10.2-1.armv7hl
	libicalvcal.so.0 is needed by (installed) nemo-qml-plugin-calendar-qt5-lightweight-0.1.4-1.armv7hl
	libicalvcal.so.0 is needed by (installed) nemo-qml-plugin-calendar-qt5-0.1.4-1.armv7hl
	libicalvcal.so.0 is needed by (installed) mkcal-qt5-0.3.27-1.16.1.armv7hl
	libicalvcal.so.0 is needed by (installed) kcalcore-qt5-4.10.2-1.armv7hl

Given that the second patch doesn't involve calls to libical, I don't think there's much point in going to that effort just yet...

Comment (pvuorela) [inline kcalcore/icaltimezones.cpp]: For testing purposes it should be enough to rebuild kcalcore on top of it (abi break) and forcing installation temporarily :)

Need to investigate myself these more, but I have a feeling they mostly work around problems, in which case I'd rather have root cause and our packages upgraded. Changelog at least seemed to have a bunch of timezone changes.

Comment (chriadam) [inline kcalcore/icaltimezones.cpp]: I built the new libical, installed it to SDK, then rebuilt kcalcore from master and install that to SDK, then rebuilt my test app, and deployed all of the packages to device -> after running test app, results are unchanged, all three tests fail.  I then rebuilt kcalcore with just the second patch applied (ie, the change to appendDateTime) on top of the new libical, and deployed that to the device, then running the test app showed all succeed (but still the interval lower-bound is printed out as empty).

In short, it looks like the new libical doesn't affect the results; the bugs are in kcalcore, I think, not libical.

I've attempted three times now to contact the kcalcore developers (first by kde-pim mailinglist, then twice by IRC on both #kde and #kontact) to ask for their comments on this PR but have received no response.  I agree that the second patch especially seems strange, but this bug is critical and really frustrating our users...

**Comment by chriadam** [comment]
So, what's the verdict on this one?  I really think we need to fix this issue ASAP.  The second patch is the least-invasive and fixes every problem the unit-test-application exposes.  Can we merge that one?

File: kcalcore/recurrencerule.cpp (hunk 2 above)
Comment (lenggi) [inline kcalcore/recurrencerule.cpp]: The problem is not the timespec value, but the `QDateTime` value, i.e. `KDateTime.isValid()` is false, because the enclosed `QDateTime` value is not valid.
The `KDateTime` is initialized with a `QDateTime` having `Qt::LocalTime` set as its timezone information. (https://github.com/mer-packages/kcalcore/blob/master/kcalcore/kdedate/kdatetime.cpp#L809)

Therefore, the timestamp that marks the beginning of the DST period is invalid, cf. also
http://doc.qt.io/qt-5/qdatetime.html#isValid

```
qWarning() << QDateTime( QDate( 2015, 03, 29 ), QTime( 02, 00, 00 ), Qt::LocalTime).isValid();
```
is false if your system time is set to 'Europe/Berlin'.
This applies for all timezones that observe DST, e.g. the following is false if the system time is set to Australia/Sydney.

```
qWarning() << QDateTime( QDate( 2015, 10, 04 ), QTime( 02, 00, 00 ), Qt::LocalTime).isValid();
```

Unfortunately, I have currently no idea how this could be fixed. My bet would be to use UTC based times to calculate these transitions, but this is only a guess due to the lack of alternatives...

Comment (chriadam) [inline kcalcore/recurrencerule.cpp]: Thank you for that information.  I wonder if the UTC-based time could ever differ from the timespec-defined time in reality - if not, then the functional part of this (second) patch should be adequate to resolve the issue, if I'm understanding correctly.

Comment (lenggi) [inline kcalcore/recurrencerule.cpp]: `timespec.utcOffset()` on has useful information if the `SpecType` is `OffsetFromUTC`, which is probably not the case, cf. https://github.com/mer-packages/kcalcore/blob/master/kcalcore/kdedate/kdatetime.h#L322

But then again, we need to have UTC times, so we do not run into invalid times due to skipped hours when switching to DST. I have created an according pull request that tries to fix the problem one level above. Please let me know what you think about it. Thanks!

Comment (chriadam) [inline kcalcore/recurrencerule.cpp]: Thank you!  Your PR looks good to me.  Hopefully Pekka has a chance to take a look soon.

**Comment by chriadam** [comment]
This PR is obsoleted by https://github.com/mer-packages/kcalcore/pull/9/files
